ossdbtoolsservice/language/completion/mysql_completer.py

```diff
@@ -0,0 +1,378 @@
+
+    users = []
+
+    def __init__(self, smart_completion=True, logger=None, settings=None):
+        super().__init__()
+        self.smart_completion = smart_completion
+        self.logger: Logger = logger
+        settings = settings or {}
+
+        self.reserved_words = set()
+        for x in self.keywords:
+            self.reserved_words.update(x.split())
+        self.name_pattern = compile(r"^[_a-z][_a-z0-9\$]*$")
+
+        keyword_casing = settings.get('keyword_casing', 'upper').lower()
+        if keyword_casing not in ('upper', 'lower', 'auto'):
+            keyword_casing = 'auto'
+        self.keyword_casing = keyword_casing
+        self.reset_completions()
+
+    def _log(self, is_error: bool, msg: str, *args) -> None:
+        if self.logger is not None:
+            if is_error:
+                self.logger.error(msg, *args)
+            else:
+                self.logger.debug(msg, *args)
+
+    def escape_name(self, name):
+        if name and ((not self.name_pattern.match(name))
+                     or (name.upper() in self.reserved_words)
+                     or (name.upper() in self.functions)):
+            name = '`%s`' % name
+
+        return name
+
+    def escaped_names(self, names):
+        return [self.escape_name(name) for name in names]
+
+    def extend_database_names(self, databases):
+        self.databases.extend(databases)
+
+    def extend_keywords(self, additional_keywords):
+        keywords_list = list(self.keywords)
+        keywords_list.extend(additional_keywords)
+        self.keywords = tuple(keywords_list)
+        self.all_completions.update(additional_keywords)
+
+    def extend_show_items(self, show_items):
+        for show_item in show_items:
+            self.show_items.extend(show_item)
+            self.all_completions.update(show_item)
+
+    def extend_users(self, users):
+        for user in users:
+            self.users.extend(user)
+            self.all_completions.update(user)
+
+    def extend_schemata(self, schema):
+        if schema is None:
+            return
+        metadata = self.dbmetadata['tables']
+        metadata[schema] = {}
+
+        # dbmetadata.values() are the 'tables' and 'functions' dicts
+        for metadata in self.dbmetadata.values():
+            metadata[schema] = {}
+        self.all_completions.update(schema)
+
+    def extend_relations(self, data, kind):
+        """Extend metadata for tables or views
+        :param data: list of (rel_name, ) tuples
+        :param kind: either 'tables' or 'views'
+        :return:
+        """
+        # 'data' is a generator object. It can throw an exception while being
+        # consumed. This could happen if the user has launched the app without
+        # specifying a database name. This exception must be handled to prevent
+        # crashing.
+        try:
+            data = [self.escaped_names(d) for d in data]
+        except Exception:
+            data = []
```

Comment:
Can we ensure that these exceptions are getting logged

Reply:
same for all below

Reply:
done